src/jupyterlabcontroller/models/v1/external/lab.py

```diff
@@ -0,0 +1,245 @@
+        description=(
+            "Username for Lab user.  Should follow general Unix"
+            " naming conventions and therefore match the regular"
+            " expression `[a-z_][a-z0-9_-]*[$]` ."
+        ),
+    )
+    name: str = Field(
+        ...,
+        title="name",
+        example="Ribbon",
+        description=(
+            "Human-friendly display name for user.  May contain"
+            " contain spaces and capital letters; should be the"
+            " user's preferred representation of their name to"
+            " other humans."
+        ),
+    )
+    uid: int = Field(
+        ...,
+        title="uid",
+        example=1104,
+        description=(
+            "Numeric UID for user (POSIX).  32-bit unsigned integer."
+        ),
+    )
+    gid: int = Field(
+        ...,
+        title="gid",
+        example=1104,
+        description=(
+            "Numeric GID for user's primary group (POSIX).  32-bit"
+            " unsigned integer."
+        ),
+    )
+    groups: UserGroupList
+
+
+class UserQuotaQuantum(BaseModel):
+    cpu: float = Field(
+        ...,
+        title="cpu",
+        example=1.5,
+        description=(
+            "Kubernetes CPU resource, floating-point value.  cf"
+            " https://kubernetes.io/docs/tasks/configure-pod-container/"
+            "assign-cpu-resource/"
+        ),
+    )
+    memory: int = Field(
+        ...,
+        title="memory",
+        example=1073741824,
+        description=("Kubernetes memory resource in bytes."),
+    )
+
+
+class UserQuota(BaseModel):
+    limits: UserQuotaQuantum = Field(
+        ..., title="limits", description="Maximum allowed resources"
+    )
+    requests: UserQuotaQuantum = Field(
+        ..., title="requests", description="Intially-requested resources"
+    )
+
+
+class UserData(UserInfo, LabSpecification):
+    status: str = Field(
+        ...,
+        title="status",
+        example="running",
+        description=(
+            "Status of user container.  Must be one of `starting`,"
+            " `running`, `terminating`, or `failed`."
+        ),
+    )
+    pod: str = Field(
+        ...,
+        title="pod",
+        example="present",
+        description=(
+            "User pod state.  Must be one of `present` or `missing`."
+        ),
+    )
+    quotas: UserQuota
+    events: Deque[Event] = Field(
+        deque(),
+        title="events",
+        description=("Ordered queue of events for user lab creation/deletion"),
+    )
+
+    @validator("status")
+    def legal_user_status(cls, v: str) -> str:
+        if v not in lab_statuses:
+            raise ValueError(f"must be one of {lab_statuses}")
+        return v
+
+    @validator("pod")
+    def legal_pod_state(cls, v: str) -> str:
+        if v not in pod_states:
+            raise ValueError(f"must be one of {pod_states}")
+        return v
+
+    def to_components(
+        self,
+    ) -> Tuple[str, str, UserInfo, LabSpecification, UserQuota]:
```

Comment:
In this case, I'm not sure there is a good way.  It's just a way of blasting apart a UserData into its constituent parts.  I could make each one individually-extractable, or I could return it as a List rather than a Tuple, but fundamentally it's just the inverse of the classmethod from_components that assembles a UserData from its parts.

Reply:
I checked the current code on the branch and it doesn't look like this function is ever called? My confusion is that I don't understand why you would ever want to do what this function is doing. I'm not sure I understand `from_components` either. It seems to be used to load some test data? But I'm not sure I understand why the test data isn't in a format so that you can just use `parse_obj` to load it.

Reply:
But it turns out I am not using it.